### PR TITLE
Bug 1853301 - Refactor ReviewQualityCheckNavigationMiddleware to use TabsUseCases and SessionUseCases

### PR DESCRIPTION
**Pull request by Alexandru2909**
Also refactored `LinkType` class to support instances for each external link accessible from the review quality check feature

### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1853301

### CODE REVIEW

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckFooter.kt [hunk 1]
```diff
@@ -41,8 +40,10 @@ fun ReviewQualityCheckFooter(
         ),
         linkTextState = LinkTextState(
             text = poweredByLinkText,
-            url = FOOTER_LINK,
-            onClick = onLinkClick,
+            url = ReviewQualityCheckNavigationMiddleware.LinkType.PoweredBy().url,
+            onClick = {
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckFooter.kt:44]: 🔍 `onclick = onLinkClick,`
reads nicer 😄

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt [hunk 2]
```diff
@@ -5,6 +5,7 @@
 package org.mozilla.fenix.shopping.store
 
 import mozilla.components.lib.state.Action
+import org.mozilla.fenix.shopping.middleware.ReviewQualityCheckNavigationMiddleware
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt:8]: Let's not make the action depend on the middleware, we can simplify by creating different actions for different link clicks. We can then add all the link urls as constants in nav middleware which just reacts to those actions.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt [hunk 3]
```diff
@@ -42,10 +52,50 @@ class ReviewQualityCheckNavigationMiddleware(
...
                 }
+                navController.navigate(NavGraphDirections.actionGlobalBrowser(null))
             }
         }
     }
+
+    /**
+     * Types of links that can be opened from the review quality check feature.
+     */
+    sealed class LinkType {
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt:71]: I think we can simplify this by using string constants for urls. If we need to use `context` for sumo links, we can map it in the `processAction` function.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckFooter.kt (hunk 1 above)
Comment (Alexandru2909) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckFooter.kt:50]: I agree, but the `onClick` from `LinkState` accepts a `String` parameter and `onLinkClick` has none.

Comment (rahulsainani) [reply fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckFooter.kt:50]: Oh yeah you're right!

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt [hunk 4]
```diff
@@ -39,13 +44,32 @@ class ReviewQualityCheckNavigationMiddleware(
...
-            }
+        selectOrAddUseCase.invoke(
+            url = actionToUrl(
+                context = this@ReviewQualityCheckNavigationMiddleware.context,
+                action = action,
+            ),
+        )
+    }
+
+    companion object {
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt:55]: We can remove this altogether and `actionToUrl` can be a private function

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt [hunk 5]
```diff
@@ -39,13 +44,32 @@ class ReviewQualityCheckNavigationMiddleware(
...
+            context: Context,
+            action: ReviewQualityCheckAction.NavigationMiddlewareAction,
+        ) = when (action) {
+            is ReviewQualityCheckAction.OpenExplainerLearnMoreLink -> SupportUtils.getSumoURLForTopic(
+                context,
+                SupportUtils.SumoTopic.HELP,
+            )
+
+            is ReviewQualityCheckAction.OpenPoweredByLink ->
+                "www.fakespot.com"
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt:72]: Let's extract this as a constant.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt [hunk 6]
```diff
@@ -85,7 +85,12 @@ sealed interface ReviewQualityCheckAction : Action {
     object ReanalyzeProduct : NetworkAction, UpdateAction
 
     /**
-     * Triggered when opening a link from the review quality check feature.
+     * Triggered when opening the learn more link from the explainer card.
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt:88]:
```suggestion
     * Triggered when user clicks on learn more link on the opt in card.
```

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt [hunk 7]
```diff
@@ -85,7 +85,12 @@ sealed interface ReviewQualityCheckAction : Action {
...
      */
-    data class OpenLink(val link: ReviewQualityCheckState.LinkType) : NavigationMiddlewareAction
+    object OpenExplainerLearnMoreLink : NavigationMiddlewareAction
+
+    /**
+     * Triggered when opening the "Powered by" link from the footer.
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/store/ReviewQualityCheckAction.kt:93]:
```suggestion
     * Triggered when the user clicks on the "Powered by" link in the footer.
```

File: fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddlewareTest.kt [hunk 8]
```diff
@@ -1,75 +1,57 @@
...
     private val dispatcher = coroutinesTestRule.testDispatcher
     private val scope = coroutinesTestRule.scope
+    private lateinit var store: ReviewQualityCheckStore
 
-    @Test
-    fun `WHEN opening an external link THEN the link should be opened in a new tab`() {
-        var isOpenedInSelectedTab = false
-        var isOpenedInNewTab = false
-        val store = ReviewQualityCheckStore(
+    @MockK(relaxed = true)
```
Comment (rahulsainani) [inline fenix/app/src/test/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddlewareTest.kt:28]: ❓ Do we need to mock this? We could use the real object using a `BrowserStore` object and assert the store state that the url is opened in a tab.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt [hunk 9]
```diff
@@ -4,21 +4,29 @@
...
 import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.launch
+import mozilla.components.feature.tabs.TabsUseCases
 import mozilla.components.lib.state.Middleware
 import mozilla.components.lib.state.MiddlewareContext
+import org.mozilla.fenix.settings.SupportUtils
 import org.mozilla.fenix.shopping.store.ReviewQualityCheckAction
 import org.mozilla.fenix.shopping.store.ReviewQualityCheckState
 
+private const val PLACEHOLDER_URL = "www.fakespot.com"
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt:18]: Think we can call the constant "POWERED_BY_URL", even if the url changes, we don't need to change the code then, only this link.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt [hunk 10]
```diff
@@ -39,13 +47,27 @@ class ReviewQualityCheckNavigationMiddleware(
...
     ) = scope.launch {
-        when (action) {
-            is ReviewQualityCheckAction.OpenLink -> {
-                when (action.link) {
-                    is ReviewQualityCheckState.LinkType.ExternalLink -> openLink(action.link.url, true)
-                    is ReviewQualityCheckState.LinkType.AnalyzeLink -> openLink(action.link.url, false)
-                }
-            }
-        }
+        selectOrAddUseCase.invoke(
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt:50]: 🔍 Can this be in one line?

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt [hunk 11]
```diff
@@ -39,13 +47,27 @@ class ReviewQualityCheckNavigationMiddleware(
...
+        )
+    }
+
+    /**
+     * Used to find the corresponding url to the open link action.
+     *
+     * @param action Used to find the corresponding url.
+     */
+    @VisibleForTesting
+    internal fun actionToUrl(
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/middleware/ReviewQualityCheckNavigationMiddleware.kt:63]: I think we can remove  @VisibleForTesting and assert on the browser state, we are asserting for the tab size, we can also assert for the tab url based on the action, and we can make this a private function. 

This way we don't test the implementation detail but the output. So in the future if this class is refactored without any behavior change, tests won't need to be updated and will help in refactoring confidently.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckBottomSheet.kt [hunk 12]
```diff
@@ -85,19 +73,17 @@ fun ReviewQualityCheckBottomSheet(
...
                     },
-                    onReviewGradeLearnMoreClick = { url ->
+                    onReviewGradeLearnMoreClick = {
                         store.dispatch(
-                            ReviewQualityCheckAction.OpenLink(
-                                ReviewQualityCheckState.LinkType.ExternalLink(url),
-                            ),
+                            ReviewQualityCheckAction.OpenExplainerLearnMoreLink,
                         )
+                        onRequestDismiss()
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckBottomSheet.kt:80]:
```
onRequestDismiss()
store.dispatch(ReviewQualityCheckAction.OpenExplainerLearnMoreLink)
```

We want to dismiss the sheet first and then navigate to a new tab, Same for action below.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckBottomSheet.kt [hunk 13]
```diff
@@ -47,26 +47,14 @@ fun ReviewQualityCheckBottomSheet(
...
+                        store.dispatch(ReviewQualityCheckAction.OpenOnboardingPrivacyPolicyLink)
                     },
-                    onTermsOfUseClick = { url ->
-                        store.dispatch(
-                            ReviewQualityCheckAction.OpenLink(
-                                ReviewQualityCheckState.LinkType.ExternalLink(url),
-                            ),
-                        )
+                    onTermsOfUseClick = {
+                        store.dispatch(ReviewQualityCheckAction.OpenOnboardingTermsLink)
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckBottomSheet.kt:57]: onRequestDismiss() missing before the action.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckBottomSheet.kt [hunk 14]
```diff
@@ -47,26 +47,14 @@ fun ReviewQualityCheckBottomSheet(
...
                         store.dispatch(ReviewQualityCheckAction.OptIn)
                     },
-                    onLearnMoreClick = { url ->
-                        store.dispatch(
-                            ReviewQualityCheckAction.OpenLink(
-                                ReviewQualityCheckState.LinkType.ExternalLink(url),
-                            ),
-                        )
+                    onLearnMoreClick = {
+                        store.dispatch(ReviewQualityCheckAction.OpenOnboardingLearnMoreLink)
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckBottomSheet.kt:51]: onRequestDismiss() missing before the action.

File: fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckBottomSheet.kt [hunk 15]
```diff
@@ -47,26 +47,14 @@ fun ReviewQualityCheckBottomSheet(
                     onPrimaryButtonClick = {
                         store.dispatch(ReviewQualityCheckAction.OptIn)
                     },
-                    onLearnMoreClick = { url ->
```
Comment (rahulsainani) [inline fenix/app/src/main/java/org/mozilla/fenix/shopping/ui/ReviewQualityCheckBottomSheet.kt:50]: I can't comment on the line but I think we can remove `LongMethod` suppression now

**Review by rahulsainani — APPROVED** [review]
Looks good! 🚀

**Pull request merged**